src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/modelchecking/ModelCheckingStrategy.kt

```diff
@@ -256,28 +313,37 @@ internal class ModelCheckingStrategy(
                 switchableThreads(iThread).random(interleavingFinishingRandom)
             }
 
-        fun isSwitchPosition() = executionPosition in switchPositions
+        fun isSwitchPosition() = currentSwitchCrashChoice == 0 && executionPosition in switchPositions
+        fun isCrashPosition() = currentSwitchCrashChoice == 1 && executionPosition in crashPositions
 
         /**
-         * Creates a new execution position that corresponds to the current switch point.
+         * Creates a new execution position that corresponds to the current switch/crash point.
          * Unlike switch points, the execution position is just a gradually increasing counter
          * which helps to distinguish different switch points.
          */
-        fun newExecutionPosition(iThread: Int) {
+        private fun newExecutionPosition(iThread: Int, isSwitch: Boolean) {
+            if (exploreCrashNode == isSwitch) return
             executionPosition++
-            if (executionPosition > switchPositions.lastOrNull() ?: -1) {
+            if (executionPosition > positions.lastOrNull() ?: -1) {
                 // Add a new thread choosing node corresponding to the switch at the current execution position.
                 lastNotInitializedNodeChoices?.add(Choice(ThreadChoosingNode(switchableThreads(iThread)), executionPosition))
```

Comment:
Is it the intended logic that immediately after every crash there is a thread switch?

Reply:
No, I just want to unify the creation of crash/switch choices. The plan here was that the position array corresponds to lastNotInitializedNodeChoices of the same type(crash or switch).

Reply:
I'll add the logic of immediate switch. I had a hope that Lincheck can deal with it.